{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python2.7/dist-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "  \"This module will be removed in 0.20.\", DeprecationWarning)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os as  sys\n",
    "from sklearn import cross_validation, metrics, preprocessing \n",
    "from sklearn.metrics import roc_curve, auc , precision_recall_curve , precision_recall_fscore_support, precision_score\n",
    "from nimfa.utils.linalg import *\n",
    "from collections import defaultdict\n",
    "from os.path import dirname\n",
    "from os.path import join\n",
    "from skfusion import fusion as skf\n",
    "from skfusion.fusion import ObjectType, Relation, FusionGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def residuals(V, V_hat):\n",
    "        \"\"\"\n",
    "        Return residuals matrix between the target matrix and its NMF estimate.\n",
    "        \"\"\"\n",
    "        \n",
    "        return(V - V_hat)\n",
    "\n",
    "def rss(V,V_hat):\n",
    "        \"\"\"\n",
    "        Compute Residual Sum of Squares (RSS) between NMF estimate and\n",
    "        target matrix [Hutchins2008]_.\n",
    "        \"\"\"        \n",
    "        X = residuals(V, V_hat)\n",
    "        return multiply(X, X).sum()\n",
    "\n",
    "def evar(target, _target):\n",
    "        \"\"\"\n",
    "        Compute the explained variance of the NMF estimate of the target matrix.\n",
    "        \"\"\"       \n",
    "        V = target\n",
    "        V_hat=_target\n",
    "        return 1. - rss(V,V_hat) / multiply(V, V).sum()\n",
    "def DiffusionKernel(AdjMatrix):\n",
    "    # 1.Computes Degree matrix  - diagonal matrix with diagonal entries = raw sums of adjacency matrix \n",
    "    DegreeMatrix = np.diag(np.sum(AdjMatrix, axis=1))\n",
    "    # 2. Computes negative Laplacian H = AdjMatrix - DegreeMatrix\n",
    "#     H = np.subtract(AdjMatrix, DegreeMatrix)\n",
    "    H = np.subtract(AdjMatrix,DegreeMatrix)\n",
    "    # 3. Computes matrix exponential: exp(beta*H)\n",
    "    K = sp.linalg.expm(0.2*H)\n",
    "    return K\n",
    "def load_ppi_data (GRNData,ppi_file):\n",
    "    \"\"\"\n",
    "    Read protein protein interaction data and convert it into Adjacency matrix\n",
    "    \n",
    "    return : Adjacency of PPI\n",
    "    \"\"\"\n",
    "    ppi=np.loadtxt(ppi_file, dtype=np.int32)\n",
    "    PPI=np.zeros(((GRNData['nvertices']),(GRNData['nvertices'])))    \n",
    "    for i in range(0,ppi.shape[0]):\n",
    "         PPI[ppi[i][0],ppi[i][1]]=1\n",
    "    return PPI\n",
    "def load_tf_tfi_data (GRNData,tf_tfi_file):\n",
    "    \"\"\"\n",
    "    Read  interaction data between known transcription factors in this data \n",
    "    \n",
    "    return : Adjacency of TF-TFI\n",
    "    \"\"\"\n",
    "    tf_tfi=np.loadtxt(tf_tfi_file, dtype=np.int32)\n",
    "    TF_TFI=np.zeros(((GRNData['nsources']),(GRNData['nsources'])))    \n",
    "    for i in range (0,tf_tfi.shape[0]):\n",
    "         TF_TFI[tf_tfi[i,0], tf_tfi[i,1]]=1\n",
    "            \n",
    "    return TF_TFI\n",
    "def load_tf_gene_interaction_data(GRNData):\n",
    "    \n",
    "    \"\"\"\n",
    "     Generate an adjacency matrix from the known regulations of E.coli benchmark data\n",
    "    \n",
    "    return : Adjacency of TF-Gene interactions\n",
    "    \"\"\"\n",
    "    \n",
    "    i,j,sourceIndices = np.unique(GRNData['edges'][:,0], return_index=True, return_inverse=True)\n",
    "    row_count,col_count=[GRNData['nvertices'], GRNData['nsources']]\n",
    "    Gold=np.zeros([row_count,col_count])\n",
    "    edges=GRNData['edges']\n",
    "    for i in range(0,GRNData['nedges']):\n",
    "        Gold[edges[i,1],sourceIndices[i]] = 1 \n",
    "        \n",
    "    return Gold\n",
    "\n",
    "def load_ecoli_data ():\n",
    "   \n",
    "    \"\"\"\n",
    "    Read   all E.coli genomic data and build a data structure GRNData for further processing.\n",
    "    \n",
    "    return :  Dictionary of Ecoli datasets and other attributes\n",
    "    \n",
    "    \"\"\"\n",
    "    DATA_DIR=\"./data/\"\n",
    "    expr_file=DATA_DIR + 'expression.txt'\n",
    "    expr_file='/home/csserver/sirene-1.1/data/ecoli/ecoli_expression_data'\n",
    "    reg_file=DATA_DIR + 'regulation.txt'    \n",
    "    go_sem_file =DATA_DIR + 'EcK12SemSim_4345.txt'\n",
    "    pp_interaction_file=DATA_DIR +'ppi_interaction_strings.txt'\n",
    "    tf_tfi_file=DATA_DIR +'tf_tf_interaction.txt'\n",
    "    GRNData=buildRegulationData(expr_file,reg_file,pp_interaction_file,go_sem_file,tf_tfi_file)\n",
    "    \n",
    "    return GRNData\n",
    "def get_genes():\n",
    "    \"\"\"\n",
    "    Read   Complete list of genes for E. coli benchmark.\n",
    "    \n",
    "    return :  List of genes.\n",
    "    \n",
    "    \"\"\"\n",
    "    DATA_DIR=\"./data/\"\n",
    "    genes=np.genfromtxt(DATA_DIR + \"gene_names.txt\",dtype='object')\n",
    "    \n",
    "    \n",
    "    return genes\n",
    "def matrix_cmpletion (fuser)\n",
    "    \n",
    "    \"\"\"\n",
    "    fuser object containing relations and their latent factors.\n",
    "    \n",
    "    return :  Reconstructed   (R12_hat) from  low rank matrix  facors\n",
    "    \n",
    "    \"\"\"\n",
    "    S12=fuser.backbones_[gene_tf_relation][i]\n",
    "    G1=fuser.factors_[gene][i]\n",
    "    G2=fuser.factors_[tf][i]\n",
    "    target_hat=np.dot(G1, np.dot(S12, G2.T))  \n",
    "    \n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def  buildRegulationData (expr_file,regulation_file,ppi_file,go_sem_file,tf_tfi_file):\n",
    "    \n",
    "     \"\"\"\n",
    "        This function reads required data files to build a datastructure containing all neceaasy information for GRN inference\n",
    "    \n",
    "        return :  Reconstructed   (R12_hat) from  low rank matrix  facors\n",
    "    \n",
    "    \"\"\"\n",
    "        \n",
    "    df_exp = pd.read_csv(expr_file,sep='\\t', header=None,index_col=None)\n",
    "    exp_matrix = df_exp.as_matrix()\n",
    "    df_regulation=pd.DataFrame.from_csv(regulation_file,sep='\\t', header=None,index_col=None)\n",
    "    df_regulation_matrix=df_regulation.as_matrix()\n",
    "    go_sem=pd.read_csv(go_sem_file, sep='\\t',header=None,index_col=None)\n",
    "    go_sem=go_sem.as_matrix()\n",
    "    tf_tfi=np.loadtxt(tf_tfi_file, dtype=np.int32)   \n",
    "    edges=df_regulation_matrix    \n",
    "    edges=edges-1\n",
    "    knownTargets= np.unique(edges[:,1])\n",
    "    nrows, ncols = exp_matrix.shape    \n",
    "    toPredict = np.setdiff1d(np.arange(nrows),knownTargets)\n",
    "    sources = np.unique(edges[:,0])\n",
    "    nsource = sources.size\n",
    "    nvertices , nfeatures= exp_matrix.shape\n",
    "    GRNData = { 'features'  :    exp_matrix , \n",
    "                'edges'     :    edges,\n",
    "                 'go_sem'   :    go_sem,                 \n",
    "                'nvertices' :    nvertices,\n",
    "                'nfeatures' :     nfeatures,                                 \n",
    "                'nedges' :       edges.shape[0],\n",
    "                'knownTargets' : knownTargets,   \n",
    "                'toPredict' :    toPredict,                 \n",
    "                'sources'   :    sources,\n",
    "                'nsources' :     nsource                    \n",
    "          }\n",
    "    GRNData['Gold']=load_tf_gene_interaction_data(GRNData)\n",
    "    \"\"\"\n",
    "        Compution of  diffusion kernel commneted out becuase of huge CPU time it takes\n",
    "    \n",
    "    \"\"\"\n",
    "#     GRNData['ppi']=DiffusionKernel(load_ppi_data(GRNData, ppi_file))\n",
    "#     GRNData['tf_tfi']=DiffusionKernel(load_tf_tfi_data(GRNData,tf_tfi_file))\n",
    "    GRNData['ppi']=load_ppi_data(GRNData, ppi_file)\n",
    "    GRNData['tf_tfi']=load_tf_tfi_data(GRNData,tf_tfi_file)\n",
    "\n",
    "    return  GRNData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main (GRNData):\n",
    "    gene=skf.ObjectType(\"Gene\",250)\n",
    "    go=skf.ObjectType(\"GO\",200)\n",
    "    exp_cond=skf.ObjectType(\"Experimental condition\",15)\n",
    "    tf=skf.ObjectType(\"TF\",15)\n",
    "    n_folds = 5\n",
    "    n_genes = len(get_genes())\n",
    "    print(n_genes)\n",
    "    cv = cross_validation.KFold(n_genes, n_folds=n_folds)\n",
    "    fold_mse = np.zeros(n_folds)\n",
    "    precision_avg=np.zeros(n_folds)\n",
    "    recall_avg=np.zeros(n_folds)\n",
    "    geneTF_mask = np.zeros_like(GRNData['Gold']).astype('bool')\n",
    "    print(geneTF_mask.shape)\n",
    "    knownTargets=GRNData['knownTargets']\n",
    "    \n",
    "    relations = [\n",
    "            skf.Relation(GRNData['Gold'],gene,tf),           \n",
    "            skf.Relation(GRNData['go_sem'],gene,go),\n",
    "            skf.Relation(GRNData['features'],gene,exp_cond),\n",
    "             skf.Relation(GRNData['ppi'],gene,gene),\n",
    "            skf.Relation(GRNData['tf_tfi'],tf,tf)                        \n",
    "    ]\n",
    "    fusion_graph = skf.FusionGraph(relations)\n",
    "    fuser = skf.Dfmc(max_iter=100, n_run=1, init_type=\"random_vcol\", random_state=0)\n",
    "    gold=0\n",
    "    for i, (train_idx, test_idx) in enumerate(cv):\n",
    "        geneTF_mask[:] = False\n",
    "        geneTF_mask[test_idx, :] = True\n",
    "        fusion_graph[gene][tf][0].mask = geneTF_mask\n",
    "        fuser.fuse(fusion_graph)\n",
    "        pred_ann = fuser.complete(fuser.fusion_graph[gene][tf][0])[test_idx]        \n",
    "        true_ann = GRNData['Gold'][test_idx]        \n",
    "        threshold=0.152415\n",
    "        print (np.sum(true_ann>0), np.sum(pred_ann>threshold))\n",
    "        C = np.where(pred_ann >= threshold, 1, 0)      \n",
    "        print(metrics.confusion_matrix(true_ann.flatten(),C.flatten()))\n",
    "        print(\"F1 score\")\n",
    "        print (metrics.f1_score(true_ann.flatten(),C.flatten()))\n",
    "        print(\"Precision score\")\n",
    "        print (metrics.precision_score(true_ann.flatten(),C.flatten()))\n",
    "        print(\"Recall Score\")\n",
    "        print (metrics.recall_score(true_ann.flatten(),C.flatten()))\n",
    "        fold_mse[i] = metrics.mean_squared_error(pred_ann, true_ann)\n",
    "        precision_avg[i]=metrics.precision_score(true_ann.flatten(),C.flatten())\n",
    "        recall_avg[i]=metrics.recall_score(true_ann.flatten(),C.flatten())\n",
    "        gold=gold+np.sum(true_ann>0)\n",
    "    pred_ann = fuser.complete(fuser.fusion_graph[gene][tf][0])\n",
    "    C = np.where(pred_ann[knownTargets] >= threshold, 1, 0)\n",
    "    print(\"SCore for Known Targets\")\n",
    "    print (np.sum(pred_ann>=threshold))\n",
    "    print (metrics.confusion_matrix(GRNData['Gold'][knownTargets].flatten(),C.flatten()))\n",
    "    print(\"Average Precision %5.4f\" % np.mean(precision_avg))\n",
    "    print (\"Average Recall %5.4f\" % np.mean( recall_avg))\n",
    "    print (\"Average F1 SCore %5.4f\" % (2*(np.mean(precision_avg)*np.mean( recall_avg))/np.mean(precision_avg)+np.mean( recall_avg)))\n",
    "    print(\"MSE: %5.4f\" % np.mean(fold_mse))     \n",
    "    print (\"Total True interactions %3.2f\" % gold)    \n",
    "    pred_ann = fuser.complete(fuser.fusion_graph[gene][tf][0])\n",
    "    return pred_ann,true_ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4345\n",
      "(4345, 154)\n",
      "(460, 501)\n",
      "[[133216    150]\n",
      " [   109    351]]\n",
      "F1 score\n",
      "0.730489073881\n",
      "Precision score\n",
      "0.700598802395\n",
      "Recall Score\n",
      "0.763043478261\n",
      "(1409, 1399)\n",
      "[[132114    303]\n",
      " [   313   1096]]\n",
      "F1 score\n",
      "0.780626780627\n",
      "Precision score\n",
      "0.783416726233\n",
      "Recall Score\n",
      "0.777856635912\n",
      "(1078, 1132)\n",
      "[[132483    265]\n",
      " [   211    867]]\n",
      "F1 score\n",
      "0.784615384615\n",
      "Precision score\n",
      "0.765901060071\n",
      "Recall Score\n",
      "0.80426716141\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "  \n",
    "       \n",
    "    ecoli_data=load_ecoli_data()      \n",
    "    predicted,gold=main(ecoli_data)    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[False False False ..., False False False]\n",
      " [False False False ..., False False False]\n",
      " [False False False ..., False False False]\n",
      " ..., \n",
      " [False False False ..., False False False]\n",
      " [False False False ..., False False False]\n",
      " [False False False ..., False False False]]\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
